sklearn/linear_model/ridge.py

```diff
@@ -1324,17 +1326,21 @@ def _solve_eigen_covariance_intercept(
         return (1 - hat_diag) / alpha, (y - y_hat) / alpha
 
     def _solve_eigen_covariance(
-            self, alpha, y, sqrt_sw, X_mean, s, V, X):
-        """Compute dual coefficients and diagonal of (Identity - Hat_matrix)
+            self, alpha, y, sqrt_sw, X_mean, eigvals, Q, X):
+        """Compute dual coefficients and diagonal of (Identity - Hat_matrix),
+        where Hat_matrix = X(X^TX + alpha*I)^(-1)X^T
 
         Used when we have a decomposition of X^T.X
-        (n_features < n_samples and X is sparse).
+        (n_samples > n_features and X is sparse).
+
+        Letting X^T.X = QVQ^T, the Hat_matrix can be written as:
+        QV(V + alpha*I)^(-1)VQ
         """
         if self.fit_intercept:
             return self._solve_eigen_covariance_intercept(
-                alpha, y, sqrt_sw, X_mean, s, V, X)
+                alpha, y, sqrt_sw, X_mean, eigvals, Q, X)
         return self._solve_eigen_covariance_no_intercept(
-            alpha, y, sqrt_sw, X_mean, s, V, X)
+            alpha, y, sqrt_sw, X_mean, eigvals, Q, X)
 
     def _svd_decompose_design_matrix(self, X, y, sqrt_sw):
```

Comment:
should the notations used in this function (especially `v = s **2`) be updated as well?

Reply:
Using the covariance matrix is left as a side note, in section 5.3. For a future reader, seeing `S^2` and looking at the reference, would make it seem like we did SVD on `X`, when in fact we did eigenvalue decomposition on `X^TX`.

Reply:
in this function we are really computing an SVD of X. maybe `v` should be called `eigvals` since it is the name we are using elsewhere

Reply:
My understanding is, in `_eigen_decompose_covariance` we are doing an eigenvalue decomposition of `X^TX` and `_solve_eigen_covariance` uses this to get 

```
X(X^TX + alpha*I)^(-1)X^T
```

Letting `X^TX=QLQ^T`, we write this as 

```
XQ^T(L + alpha*I)^(-1)QX^T
```

I may be missing something. Where are we doing SVD directly in this procedure?

Reply:
you are right. In what you write above I would only change one detail: in the reference Q and L are the eigenvectors and eigenvalues of XX^T, whereas V and S^2 are the eigenvectors and eigenvalues of X^TX. (also Q is transposed in the second equation). So to avoid confusion I would replace 

> Letting `X^TX=QLQ^T`, we write this as
> 
> ```
> XQ^T(L + alpha*I)^(-1)QX^T
> ```

with 
Letting `X^TX=VS^2V^T`, we write this as

```
XV(S^2 + alpha*I)^(-1)V^TX^T
```



> Where are we doing SVD directly in this procedure?

We are not doing SVD in `_eigen_decompose_covariance`, but in `_svd_decompose_design_matrix` (when `X` is dense we compute `U` in the decomposition to avoid having to recompute the product `XV` many times)

Reply:
> (also Q is transposed in the second equation).

Oops.

> Letting X^TX=VS^2V^T, we write this as

I think adding any reference to SVD would be confusing for a future reviewer. We are following section 5.3 starting from 

```
1-X(X^TX+alpha*I)^-1X^T
```

and not following the reference using SVD in section 5.2.